luajit/upipe.lua

```diff
@@ -365,6 +365,28 @@ ffi.metatype("struct uref", {
     end
 })
 
+--[[
+-- A function to iterate over all the planes of a uref.  For each plane it
+-- returns the plane string (chroma for pic, channel for sound).
+--
+-- The value for the plane will be invalid on the next iteration so make a copy
+-- if it needs to be kept.
+--]]
+local function foreach_plane(func)
+    return function(ref)
+        local plane = ffi.new("const uint8_t *[1]")
+        return function()
+            if C.ubase_check(func(ref, plane)) and plane[0] ~= nil then
+                return plane[0]
+            else
+                return nil
+            end
+        end
+    end
+end
+uref_pic_foreach_plane = foreach_plane(C.uref_pic_iterate_plane)
+uref_sound_foreach_plane = foreach_plane(C.uref_sound_iterate_plane)
+
```

Comment:
I try to avoid global functions. What about including them in the metatypes instead:

```diff
--- a/luajit/upipe.lua
+++ b/luajit/upipe.lua
@@ -359,7 +359,10 @@ ffi.metatype("struct upipe", {
 })
 
 ffi.metatype("struct uref", {
-    __index = function (_, key)
+    __index = function (ref, key)
+        if key == 'pic_foreach_plane' then
+            return foreach_plane(ref, C.uref_pic_iterate_plane)
+        end
         local f = C[fmt("uref_%s", key)]
         return getter(uref_getters, f, key) or f
     end
```

Reply:
That would make it more Lua-like

I will try both changes and make additional commits if they work.